### ISSUE TITLE
Problem rendering HTML tag attributes

### ISSUE THREAD

**Issue by ascastro** [YorickPeterse/oga#142]
We came across the following issue while parsing HTML. The following code:
```ruby
Oga::parse_html('<a class="addthis_button_google_plusone_badge" g:plusone:size="small"></a>').to_xml
```
would result in an exception:
`
LL::ParserError: Unexpected attribute namespace, expected attribute instead on line 1
`

Would it be possible to handle those attributes?

Thanks!

**Comment by YorickPeterse** [comment]
I wasn't aware this is even valid HTML (is it?). What is supposed to be the attribute name here, `size`, `plusone:size`, or `g:plusone:size`? And is this valid in XML, or HTML only?

**Comment by ascastro** [comment]
Thanks for replying so quickly.

There are lot of web pages using that syntax to generate Google +1 buttons, Facebook, ... I believe this comes from AddThis, but I can't be sure if it's the only place:
http://support.addthis.com/customer/portal/articles/381237-third-party-button

The attribute name would be `size`, and the element `g:plusone`.

Another option would be to have a 'non-strict' mode when parsing HTML, so that we could swallow those exceptions and skip attributes that can't be parsed.

**Comment by omninonsense** [comment]
Browsers would treat this attribute simply as `g:plusone:size` and allow it, but it's technically not valid HTML5 (according to [validator.w3.org](https://validator.w3.org/)):

~~~javascript
var div, a;

div = document.createElement('div');
div.innerHTML = '<a class="addthis_button_google_plusone_badge" g:plusone:size="small"></a>';

a = div.children[0];
a.getAttribute("g:plusone:size"); // "small"
~~~

**Comment by YorickPeterse** [comment]
There are two options that I can think of:

1. When parsing in HTML mode we simply treat any number of colon separated identifiers as a single name. Thus for `foo:bar="10"` the attribute name is `foo:bar`, while for `foo:bar:baz="10"` the attribute name is `foo:bar:baz`.
2. Option 1 is _only_ applied if more than 1 colon is present, meaning that `foo:bar="10"` results in the attribute name being `bar` and the namespace prefix being `foo` (this is how Oga currently behaves), while `foo:bar:baz="10"` means the attribute name is `bar:baz`.

Since namespaces are pretty much invalid/unused in HTML I'm leaning towards option 1 (this would also match libxml behaviour). This would probably require some tweaking in the lexer so it emits a single `T_ATTR` node containing the whole name, instead of also emitting `T_ATTR_NS` nodes.

**Comment by ascastro** [comment]
+1 for option 1!

**Comment by omninonsense** [comment]
+1 for Opt 1, too.

**Comment by YorickPeterse** [comment]
Namespace wise this should be backwards compatible as namespaces are already disabled in HTML mode (e.g. `Oga::XML::Element#available_namespaces` always returns an empty Hash for HTML documents). Attribute naming however this would not be backwards compatible. For example, in current versions `<foo:bar />` results in the attribute name being "bar". Fixing this particular issue would mean the attribute name would instead be "foo:bar", which might break existing code. I have no particular issues with bumping the major version (that's what semantic versioning is for), but I'll give it some extra thought over the coming days before changing anything.

**Comment by abotalov** [comment]
[The following info from HTML spec seems to be relevant](http://www.w3.org/TR/html51/syntax.html#syntax-attribute-name):
> Attribute names must consist of one or more characters other than the space characters, U+0000 NULL, U+0022 QUOTATION MARK ("), U+0027 APOSTROPHE ('), ">" (U+003E), "/" (U+002F), and "=" (U+003D) characters, the control characters, and any characters that are not defined by Unicode

See also http://www.w3.org/TR/html51/syntax.html#attribute-name-state

**Comment by YorickPeterse** [comment]
Forgot to mention this: the fix for this has been released in 2.0.
